esiclient/tests/functional/base.py

```diff
@@ -0,0 +1,225 @@
+                user_domain_name=self.users[user]['domain'],
+                tenant_name=self.projects[project]['name'],
+                project_domain_name=self.projects[project]['domain'],
+                identity_api_version='3',
+                uri=self.config['auth_url']))
+
+    def _init_dummy_project(self, name, roles, parent=None):
+        admin_client = self.clients['admin']
+
+        # NOTE: names created using tempest's data_utils.rand_name() function
+        #       will be prefixed with 'tempest-' and suffixed with randomly
+        #       generated characters to prevent name collision.
+        project_name = data_utils.rand_name('esi-%s-project' % name)
+        flags = '--domain default --enable -f json'
+        if parent:
+            if parent not in self.projects.keys():
+                raise NameError('Invalid parent project: %s' % parent)
+            flags += ' --parent %s' % self.projects[parent]['name']
+        output = admin_client.openstack('project create %s' % project_name,
+                                        '', flags)
+        self._cls_cleanups.append([admin_client.openstack, 'project delete',
+                                   '', '%s' % project_name])
+        project_id = json.loads(output)['id']
+        self.projects[name] = {
+            'name': project_name,
+            'id': project_id,
+            'parent': parent,
+            'domain': 'default'
+        }
+
+        if roles is []:
+            raise ValueError('No roles specified when initializing dummy \
+                              project %s' % name)
+        elif type(roles) is str:
+            roles = [roles]
+
+        for role in roles:
+
+            username = data_utils.rand_name('esi-%s-%s' % (name, role))
+            password = data_utils.rand_password()
+            output = admin_client.openstack('user create %s' % username, '',
+                                            '--domain default --password %s \
+                                             --enable -f json' % password)
+            self._cls_cleanups.append([admin_client.openstack, 'user delete',
+                                       '', '%s' % username])
+            user_id = json.loads(output)['id']
+
+            client_name = '%s-%s' % (name, role)
+            self.users[client_name] = {
+                'name': username,
+                'id': user_id,
+                'password': password,
+                'domain': 'default'
+            }
+
+            admin_client.openstack('role add', '', '--user %s --project %s %s'
+                                   % (username, project_name, role))
+            self._cls_cleanups.append([admin_client.openstack, 'role remove',
+                                       '', '--user %s --project %s %s' %
+                                       (username, project_name, role)])
+
+            self._init_client(self, client_name, name)
+
+    def _init_random_client(self):
```

Comment:
It isnt quite as easy to do as you think, because of the weirdness with the class and instance method stuff in the base of the code, but I suppose it can be done.

Reply:
What error do you get if you do what I suggest? Anyway, I'd recommend trying to get that to work because the amount of duplicated code is probably a bad sign.